### ISSUE TITLE
MSL AD gate (#530): open question — does the band-mean |S21|^2 gradient reflect beta/reflection physics or a frozen z0_hj reference-plane artifact?

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#560]
## Open question, split out of PR #559 (issues #530/#515)

PR #559 replaced the MSL AD gate's objective (`test_msl_ad_fd_converged_tight`)
with band-mean `|S21|**2` and rebuilt the #515 AD smoke, both now sharing
`tests/_msl_ad_objective.py::msl_band_mean_s21_sq`. An adversarial review of
that PR caught an earlier draft claiming the new objective "moves directly
with... guided wavelength via beta" — that mechanism claim is **unmeasured**,
and this issue tracks the decisive probe that would settle it, which PR #559
ships without running (documented as open, not run, in three places — see
"Where this is currently documented" below).

## The two candidate mechanisms

`d(band-mean |S21|**2)/d(alpha)` (`alpha` scales `eps_r_sub` over the whole
grid) could be dominated by either:

1. **Genuine physics**: `alpha` shifts the guided wavelength via `beta`,
   which shifts the standing-wave pattern the line's electrical length puts
   at each port — a real beta/reflection channel.
2. **A reference-plane artifact**: `compute_msl_s_matrix`'s wave split
   computes `a = (V + Z0*I)/2`, `b = (V - Z0*I)/2` using the FROZEN analytic
   Hammerstad-Jensen `z0_hj` as the reference impedance (not the per-frequency
   FITTED `z0` the N-probe extractor also produces). As `alpha` changes the
   line's true characteristic impedance, the frozen `z0_hj` reference drifts
   further from (or closer to) the true `z0`, which can itself synthesize an
   `alpha`-dependent apparent reflection/transmission split that has nothing
   to do with beta or a genuine standing-wave shift.

## The sign witness (why mechanism 2 is the leading candidate, not yet proven)

On the gate's own fixture, `g_ad > 0` (increasing `alpha` increases band-mean
`|S21|**2`, i.e. increasing `eps_r_sub` REDUCES apparent reflection). A
reference-plane-mismatch mechanism has a natural sign preference: as `alpha`
moves the true `z0` toward (or away from) the frozen `z0_hj`, the induced
"reflection" moves monotonically with the gap, so a consistent sign across
the measured range is exactly what mechanism 2 predicts. A genuine
beta/standing-wave channel would have no particular sign preference — the
sign would depend on where in the standing-wave pattern the ports happen to
sit, not systematically favor one direction as `alpha` grows. This is
suggestive, not decisive: it has not been checked against an independent
prediction of either mechanism's expected sign.

## The decisive probe (not run — this is the ask)

Anchor the wave split (`a`, `b` in `rfx/api/_sparams.py`, the
`a_fwd_d`/`b_ref_d` computation near the driven-port block) on the
per-frequency FITTED `z0` (`res_p["z0"]`, already computed by
`extract_msl_nprobe` earlier in the same function and available as
`Z0_per_run`/`z0_fit` in that scope) instead of the frozen analytic
`z0_hj_per_port[driven]`, then re-run ONE `jax.grad` call of
`msl_band_mean_s21_sq` w.r.t. `alpha` on the gate's fixture:

- **If `|g_ad|` collapses** (drops toward the FD-unresolvable floor, i.e.
  the reference-plane mismatch was supplying most of the sensitivity): the
  gradient's dominant channel is the frozen-reference normalization gap, not
  beta/standing-wave physics. That would be worth knowing before anyone reads
  a future `eps_override` gradient on this objective as "the line's
  electrical response to substrate permittivity" — it would more precisely
  be "how far the frozen-reference wave split drifts from self-consistency."
- **If `|g_ad|` stays close to the current `1.602933e-03`** (does not
  collapse): the reference-plane channel is NOT dominant, and the
  beta/reflection-physics reading is supported (though still not fully
  proven — a self-consistent `z0_fit`-based split changes MORE than just the
  reference impedance one term at a time, so a careful read of the result is
  needed either way).

This is a single forward+backward AD call on the existing gate fixture — cheap
(same order of cost as the smoke/probe scripts already in this PR, not a
full VESSL run), doable locally on CPU at reduced scale first, then at gate
scale if the reduced-scale result is decisive.

## Where this is currently documented as OPEN (link back here once filed)

- `tests/_msl_ad_objective.py` (module docstring, "WHY BAND-MEAN..." section)
- `tests/test_msl_ad_fd_converged.py` (`test_msl_ad_fd_converged_tight`
  docstring, "WHAT DRIVES THE GRADIENT IS AN OPEN QUESTION" section)
- `docs/guides/sparameter_support_matrix.md` (the `add_msl_port(...)` /
  `compute_msl_s_matrix(...)` row's evidence bullet)
- `docs/guides/sparameter_support_matrix.json` (same row's `ad_evidence` /
  `numeric_metrics` fields — being brought to parity with the `.md` in the
  same change that files this issue)

All four currently say "open question" / "not run" without a tracker number;
this issue is that tracker. Update those four references to cite this issue
number once assigned.

## Scope note

This issue is purely about the open MECHANISM question. It does not affect
`test_msl_ad_fd_converged_tight`'s pass/fail status, its threshold (`0.03`),
or any claims-bearing number in PR #559 — the gate's job (AD reproduces the
FD of whatever the code computes) is unaffected either way; this is about
what physical story to attach to the gradient's magnitude/sign, which the
gate itself never claimed to establish.

Refs #530, #515, PR #559.

**Comment by bk-squared** [comment]
## Decisive probe run — RESOLVED

Ran the probe this issue pre-scoped: `jax.grad` of the identical
`msl_band_mean_s21_sq` objective on the gate's own fixture
(`num_periods=20, n_freqs=8`), swapping `compute_msl_s_matrix`'s wave-split
reference impedance between the frozen analytic Hammerstad-Jensen `z0_hj`
(production) and a frozen per-port FITTED `z0` (measured at `alpha=1`,
held constant). Zero production-code edits — the swap is a scoped
`unittest.mock.patch` of `hammerstad_jensen_z0_eps_eff` in a new diagnostic
script, `scripts/diagnostics/msl_ad_z0_anchor_probe.py` (full
pre-declaration of hypotheses/decision-rule/injection-point in its module
docstring). Run log: `scripts/diagnostics/msl_ad_z0_anchor_probe_run_20260806.md`.

**Measured** (CPU, float32, this exact fixture — CPU is adequate per this
issue's own text since this is a channel-attribution question, not a
precision comparison; deterministic bit-exact across repeats for both
anchors):

```
g_a (production, frozen analytic z0_hj)   = 1.602236e-03
g_b (frozen fitted z0, held at alpha=1)   = 6.885110e-05
|g_a| / |g_b|                             = 23.27
```

`g_a` on CPU matches the committed owner-platform GPU number
(`1.602933e-03`) to within 0.04%, confirming this run reproduces the
gate's own fixture. `|g_ad|` **collapses ~23.3x**, 4.6x past this issue's
own proposed 5x collapse threshold — not a borderline call.

**VERDICT**: the frozen wave-split reference (mechanism 2 — the reference-
plane normalization gap) is the DOMINANT channel behind
`d(band-mean|S21|**2)/d(alpha)` on this fixture, not genuine beta/
standing-wave physics (mechanism 1). A future `eps_override` gradient on
this objective should be read as "how far the frozen wave-split reference
sits from the line's true impedance", not as "the line's electrical
response to substrate permittivity".

**Also worth recording**: reading `compute_msl_s_matrix` end to end shows
this issue's literal injection point (the `a_fwd_d`/`b_ref_d` block near
the driven-port S11 computation) is actually a vestigial intermediate on
this 2-port fixture — it's unconditionally overwritten by the issue-#507
multi-drive `S = B @ A^-1` solve, which is what actually reaches the
objective. The correct injection point is `z0_hj_per_port`'s origin
(`hammerstad_jensen_z0_eps_eff`), which feeds every consumer (including
the multi-drive solve's wave amplitudes) self-consistently. Confirmed via
`compute_msl_s_matrix`'s own `assembly` metadata: `'multi_drive_solve'` on
this fixture, both before and after the swap.

**What this does NOT change**: `test_msl_ad_fd_converged_tight`'s
pass/fail status, its 0.03 threshold, or its validity as an AD-vs-FD
comparator — the gate differentiates the same function on both sides
regardless of which physical channel dominates the gradient's magnitude.
Only the physical interpretation attached to a passing gate's gradient
changes.

Full details, decision-rule pre-declaration, and the honest account of a
background-task duration limit killing the first combined run mid-way
(and the reconfirmation run that closed the resulting determinism gap):
PR #587.

**Comment by bk-squared** [comment]
Decisive result recorded above (ratio ~23.3x, 4.6x past the collapse threshold, deterministic bit-exact). Closing — see PR #587 for the full diff and evidence artifacts.

**Comment by bk-squared** [comment]
## Correction to the previous comment (adversarial review caught a provenance-fabrication defect)

My previous comment on this issue claimed a "5x collapse threshold" was
"proposed" by this issue. **That's wrong and I want to correct the
record explicitly.** Re-checked: `gh issue view 560 --json body,comments
-q '.body'` has zero occurrences of "5x" or any numeric ratio. This
issue's actual criterion is QUALITATIVE, quoted verbatim from the
"decisive probe" section above:

> "If `|g_ad|` collapses (drops toward the FD-unresolvable floor, i.e. the
> reference-plane mismatch was supplying most of the sensitivity) [...] If
> `|g_ad|` stays close to the current `1.602933e-03`"

The "5x" figure came from the task prompt that commissioned the probe
(not from this issue), and got mis-cited in the probe script's docstring
as "quoted verbatim from the issue body." That was a real defect — caught
in adversarial review of PR #587 — and it's now fixed throughout that PR
(script, run log, JSON, and the support-matrix docs).

**The good news: the corrected framing makes the verdict STRONGER, not
weaker.** Operationalizing this issue's own qualitative wording against
this repo's established resolving-power discipline (issue #527): the
first-order estimate of anchor B's FD signal at the gate's h is only
**~1.155 ULP** of a float32 loss — below the **4.449 ULP** #527 measured
for the retired objective's comparator and declared untrustworthy. `g_b`
is noise-floor by this repo's own established standard, which settles
this issue on its own terms, independent of any numeric ratio. The
~23.3x figure from my earlier comment is now correctly labeled as this
PR's own pre-declared secondary check (which it always was — the mistake
was only in whose words it was attributed to).

Also added to PR #587 (not previously flagged): anchor B's own loss
exceeded 1 (a passivity violation on the raw, unprojected `eps_override`
channel) — attributed and explained, and noted as evidence that
`compute_msl_s_matrix`'s production wave split anchoring on the fitted z0
instead of the frozen analytic one is a SEPARATE, undecided design
question this work does not settle.

Verdict is unchanged: **COLLAPSE, mechanism 2 (frozen reference-plane
mismatch) dominant.** Updated evidence: PR #587, head SHA `dcb4182`.
